### PR TITLE
Envoy Integration Test Windows

### PR DESCRIPTION
**Pull request by absolutelightning**
### Description

* Added support for envoy integration tests for Windows 2019.
* Added Github workflow file so that IT runs on every PR.

### Testing & Reproduction steps

CI workflows can be seen here: https://github.com/hashicorp/consul/actions/workflows/test-integration-windows.yml.  
and https://github.com/hashicorp/consul/actions/workflows/test-integrations-windows.yml
Currently passing.

### CODE REVIEW

File: .github/workflows/test-integration-windows.yml
Comment (jmurret) [inline .github/workflows/test-integration-windows.yml:4]: can we rename the file to `test-integrations-windows` to match this name as well as the existing `test-integrations.yml` that it is based off?

File: build-support-windows/BUILD-IMAGES.md
Comment (jmurret) [inline build-support-windows/BUILD-IMAGES.md:1]: can the move this directory as `/windows` underneath `./build-support` as a way to not expand the top level and embed this in the related topic of that directory for supporting build regardless of run time.

File: WINDOWS-TEST.md
Comment (jmurret) [inline WINDOWS-TEST.md:1]: can you move this file from the root to `test/integration/connect/envoy/WINDOWS-TEST.md`

**Review by jmurret — COMMENTED** [review]
I have not reviewed this in detail but had a couple of organizational suggestions.  Will review further.

**Comment by absolutelightning** [comment]
> I have not reviewed this in detail but had a couple of organizational suggestions. Will review further.

Resolved.

**Review by jmurret — CHANGES_REQUESTED** [review]
🔥 great work.  some small questions, but overall looks good. One request in the future for CI work is if the CI is cron, lets put a link in the description for the reviewer to easily verify that it has passed.  I have added that, so just something to think about next time.  Great work!  I can see this was a lot to pull together! 🙏

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml:508]: somehow this test is not passing on CI. It is however passing on my local windows setup on ec2.
https://github.com/hashicorp/consul/actions/runs/5523416412/jobs/10074530984

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml:100]: sure

Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml]: I tried but the matrix is producing more than 256 combinations - https://github.com/hashicorp/consul/actions/runs/5598185493 
So workflow is not triggering.
The reason tests are not run together in windows, is because they fail intermittently when run in groups separated by `|`. I discussed this with @huikang as well.

Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml]: What should be the next steps then @jmurret ?

Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml:102]: Can you link to the intermittent failures you experienced?  Before adding 1000 lines of CI code, we need to fully show the original approach did not work.  We should be running this just like the non-wondows tests and then documenting in comments if it does not work.  So I would love to see an example of intermittent failures that we might try to resolve before moving on to plan b, plan c, etc.  We clearly do this already, so if we take that time box a resolution to say 3 hours, we could possibly move it forward.  I'd rather not brainstorm on the other things before we close the door on the original approach that should match non-windows.

File: .github/workflows/test-integrations-windows.yml
Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml]: ok, so what is the plan then for it?  It says unable to load certificate, so we need to troubleshoot that further, but ultimately before we merge this PR, we have to have accounted for the tests and resolve anything commented out.  So are actively working on this?  Is it a backlog ticket that is a fast follow?  Until we have a plan for this, it seems like this needs to be troublshot further and resolved, no?  If your team wants to address it in a later task, then adding a comment with task ID would be a sufficient resolution to move the PR forward.

File: .github/workflows/test-integrations-windows.yml
Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml:102]: Related to my other comment, if your team wants to move this PR forward and get tests running nightly, I think a sufficient resolution would be to add a task as a fast follow to unblock this PR.

My concerns with this is that is 1k extra lines where we have to remember to add new tests to each we add a test (IE the CI job will not pick it up automatically as it does in the non-windows version).  So if we want to move this PR forward, I would expect this to be a short lived solution with a fast follow task that is meant to be worked in on the next month.

Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml:102]: Some tests were failing randomly when we ran it in my EC2 Windows machine when grouped together. When run individually they were passing. It is not on CI. Also some tests we need to skip on windows like case-wasm, case-ingress-gateway-tls (because of windows support). Please tell me how to go about this?

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml]: Not sure how to resolve this. Because this passes on my EC2 Windows machine but fails on CI. Will create a task for this.

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml:102]: Will create a task for this.

Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml:109]: you can skip this in CI by piping `grep -v <test case name to exclude>`. 

Getting back to the original question, can you try the original approach and CI and document how it performs.  I would think if it passes in CI but flakes for you locally, it would be better to commit the original passing version of the CI configuration.  So before we backlog this, can you run the original approach in CI and link the results.  The plan b option of adding 1k lines to run each test individually should be avoided if possible, IMO.  We have no guarantees on backlog tasks ever getting worked on, so before we pass on this, it would be good to show it not working in CI and that being a blocker.

File: .github/workflows/test-integrations-windows.yml
Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml]: Please add the following information to any task you create for things getting pulled:
- link to the failed run in CI
- exact error message being received
- steps you took to trouble this and the results
- hypothesis of what is going
- nest steps

We want to make sure we are not just removing and commenting out required tests based on `it did not pass`.  We should provide of a picture of that a timeboxed effort was taken to make this work and what was found.  So far, we are describing "it didn't work in CI" without any detail.  So, it does not necessarily justify pulling it as a requirement for a later date.  Basically if we are excluding thing, we should provide comments for the context of why it is not currently supported or in the case of failure, a clear picture that it was researched but it is complex beyond the original timeboxed effort.

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [inline .github/workflows/test-integrations-windows.yml:92]: @jmurret - Please see - https://github.com/hashicorp/consul/actions/runs/5608653774/jobs/10261518132
Jobs are failing when grouped together. Not sure why

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml]: fixed.

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml]: Tests running together with job matrix - https://github.com/hashicorp/consul/actions/runs/5608653774/jobs/10261517975

File: .github/workflows/test-integrations-windows.yml
Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml]: If it is fixed, can this be uncommented?

File: .github/workflows/test-integrations-windows.yml
Comment (jmurret) [reply .github/workflows/test-integrations-windows.yml:105]: I am happy to move on from this if you and your team (@3loka) are aligned that this is getting separated out and what its priority is.

I am not trying to be an obstacle on this PR.  CI PRs are very time consuming, are huge, and take 300 runs to get right.  I just went through 2 months of nothing but migrating CI from CircleCI to GitHub actions.

Some things I can tell about helping the review process go faster:
- the reviewer will not know the ins and outs of every file, so for the most part they are checking:
  - does it work?
  - are all of the intended verifications accounted for?

So, as a PR author, I want to:
- make it easy for the reviewer to find proof this all works
  - links to where to verify this
- make it easy for the reviewer to see that all verifications are included and if there is not there is a plan for them
  - that includes commenting on why something is ommitted
  - also want the plan for it is
  - also mentioning that the team discussed this and agreed on this course of action
  - reviewers should know that the author is not just excluding stuff without team buy in

Working on CI can get very tiresome, so I appreciate and am impressed with your efforts.  I would ask that you add more detail related to the above when drafting PRs.  Also, time can obviously get lost in timezone differences.  So we should always be looking for ways to reduce the back and forth.  So, from any authoring side, if you could provide more thorough explanations in responses.  For example, a lot of your responses we short "It did not work in CI".  I think providing additional context in responses like what the error is, a link to the failure you saw, what you tried and where you gave up, etc. makes it easier on the reviewer to know again that everything has been thoroughly investigated in a time boxed manner and the author is aligned with their team on moving this forward and adding follow up tasks.  I'm open to any suggestions you have on the reviewing side to help lessen the time it takes to get to a common understanding!

**Review by jmurret — APPROVED** [review]
@absolutelightning great job.  I can see the matrix thing is a slippery slope of research, so agree on the baclogging a ticket if you are aligned with your team.  There are two unresolved conversations that should be easy to resolve:
- enable the datadog upload
- uncomment the `Envoy Integration Tests for windows case-consul-exec` test if it is fixed as you mentioned.
Approving assuming you will resolve these matters before merging.  Again, nice work.  This looked like a big effort!

File: .github/workflows/test-integrations-windows.yml
Comment (absolutelightning) [reply .github/workflows/test-integrations-windows.yml]: i meant the jira task is fixed.

**Comment by absolutelightning** [comment]
> @absolutelightning great job. I can see the matrix thing is a slippery slope of research, so agree on the baclogging a ticket if you are aligned with your team. There are two unresolved conversations that should be easy to resolve:
> 
> * enable the datadog upload
> * uncomment the `Envoy Integration Tests for windows case-consul-exec` test if it is fixed as you mentioned.
>   Approving assuming you will resolve these matters before merging.  Again, nice work.  This looked like a big effort!

Data dog upload is done. Please check.

**Pull request merged**
